lib/backends/sdc/containers.js

```diff
@@ -559,60 +525,302 @@ function addRulesToFWAPI(opts, rules, payload, callback) {
+            format('Networks: listNetworks result for %s', listParams.name));
 
         if (err) {
-            callback(errors.napiErrorWrap(err, 'problem listing networks'));
+            callback(errors.napiErrorWrap(err,
+                format('Networks: problem listing network %s',
+                    listParams.name)));
+            return;
+        }
+
+        if (networks.length < 1) {
+            callback(new errors.DockerError(
+                format('unable to find %s network uuid', listParams.name)));
+            return;
+        }
+
+        payload.networks = [ { uuid: networks[0], primary: true} ];
+        return callback();
+    });
+}
+
+/*
+ * When fabrics are configured, but no specific network is supplied,
+ * or if 'bridge' is supplied, we will use the user's default fabric
+ * network, stored in UFDS.
+ *
+ * Additionally, if the user is publishing ports or we have the
+ * `fabricRequireExternal` option (currently used only by docker build,
+ * see DOCKER-705), we will also attach to configured network pool
+ * (typically the external/public pool).
+ */
+function defaultFabricNetwork(opts, container, payload, callback) {
+    var dc = opts.app.config.datacenterName;
+    var log = opts.log;
+    var requireExternal = false;
+
+    if (publishingPorts(container) || opts.fabricRequireExternal) {
+        requireExternal = true;
+    }
+
+    log.debug({ requireExternal: requireExternal },
+        'Networks: using default fabric network');
+
+    opts.app.ufds.getDcLocalConfig(opts.account.uuid, dc, function (err, conf) {
+        log.debug({err: err, conf: conf, account: opts.account.uuid},
+            'Networks: get DC local config');
+
+        if (err || !conf || !conf.defaultnetwork) {
+            callback(errors.ufdsErrorWrap(err,
+                'Networks: could not get default network'));
             return;
         }
 
-        networks.forEach(function (n) {
-            if (!external_net
-                && n.name === netName) {
+        payload.networks = [ {uuid: conf.defaultnetwork} ];
+
+        if (requireExternal) {
+            payload.networks.push({uuid: opts.config.overlay.externalPool});
+        }
+
+        payload.networks[payload.networks.length - 1].primary = true;
+        return callback();
+    });
+}
+
+/*
+ * When fabrics are configured and a specific network name is supplied,
+ * we will find and use the specified network.
+ */
+function namedNetwork(opts, container, payload, callback) {
+
+    // need to search on networks by: name, fabric-true, owner_uuid
+    var log = opts.log;
+    var query = container.HostConfig.NetworkMode;
+    var napi = getNapiClient(opts.config.napi);
 
-                external_net = n.uuid;
+    vasync.parallel({
+        funcs: [
+            function byName(cb) {
+                var listParams = {
+                    name: query,
+                    fabric: true,
+                    owner_uuid: opts.account.uuid,
+                    limit: 2
```

Comment:
Removed; was left over from a previous version searching on a partial uuid (where >1 result would have been the user underspecifying). Anyway, removed here, not appropriate.